lib/factor-rules.js

```diff
@@ -0,0 +1,373 @@
+      const l2 = thirdFactors[j]
+      // both positive
+      let one = l1[0] * l2[1] + l1[1] * l2[0]
+      let two = l1[0] * l2[0] + l1[1] * l2[1]
+      // both negative
+      let three = l1[0] * -l2[1] + l1[1] * -l2[0]
+      let four = l1[0] * -l2[0] + l1[1] * -l2[1]
+      // l2[0] is negative and l2[1] is positive
+      let five = l1[0] * l2[1] + l1[1] * -l2[0]
+      let six = l1[0] * -l2[0] + l1[1] * l2[1]
+      // l2[0] is positive and l2[1] is negative
+      let seven = l1[0] * -l2[1] + l1[1] * l2[0]
+      let eight = l1[0] * l2[0] + l1[1] * -l2[1]
+
+      if (one == secondCoef && l2[0] * l2[1] == thirdCoef){
+        combo = [[l1[0],l2[0]],[l1[1],l2[1]]]
+      } else if (two == secondCoef && l2[0] * l2[1] == thirdCoef){
+        combo = [[l1[0],l2[1]],[l1[1],l2[0]]]
+      } else if (three == secondCoef && -l2[0] * -l2[1] == thirdCoef){
+        combo = [[l1[0],-l2[0]],[l1[1],-l2[1]]]
+      } else if (four == secondCoef && -l2[0] * -l2[1] == thirdCoef){
+        combo = [[l1[0],-l2[1]],[l1[1],-l2[0]]]
+      } else if (five == secondCoef && -l2[0] * l2[1] == thirdCoef){
+        combo = [[l1[0],-l2[0]],[l1[1],l2[1]]]
+      } else if (six == secondCoef && -l2[0] * l2[1] == thirdCoef){
+        combo = [[l1[0],l2[1]],[l1[1],-l2[0]]]
+      } else if (seven == secondCoef && l2[0] * -l2[1] == thirdCoef){
+        combo = [[l1[0],l2[0]],[l1[1],-l2[1]]]
+      } else if (eight == secondCoef && l2[0] * -l2[1] == thirdCoef){
+        combo = [[l1[0],-l2[1]],[l1[1],l2[0]]]
+      } else {
+        combo = null
+      }
+    }
+  }
+  return combo
+}
+
+// TODO: handle case when exponent is a polynomial (x^b^4)
+// TODO: have a rule to order terms in a polynomial in descending order
+
+/*
+  Factoring trinomials algorithm
+  Assumes expression is in standard form ax^2 + bx + c (descending order)
+  Assumes GCD has already been factored out
+*/
+// e.g. 12x^2 + 17x + 6 -> (3 x^1 + 2) (4 x^1 + 3)
+export const FACTOR_SUM_PRODUCT_RULE = defineRule(
+  (node) => {
+    if (query.isAdd(node) && node.args.length === 3) {
+      const [firstTerm, secondTerm, thirdTerm] = node.args
+
+      // First two terms should be polynomials
+      // Third term is either number or polynomial
+      const isTrinomial =
+            isPolynomialTerm(firstTerm)
+            && isPolynomialTerm(secondTerm)
+            && (query.isNumber(thirdTerm) || isPolynomialTerm(thirdTerm))
+
+      if (isTrinomial) {
+        let isFactorable = false
+        const {constants, coefficientMap} = getCoefficientsAndConstants(node)
+        const variables = Object.keys(coefficientMap).map(key => parse(key))
+        const coeffs = Object.keys(coefficientMap).map(key => coefficientMap[key][0])
+
+        const firstPoly = query.isMul(variables[0]) ? variables[0].args : [variables[0]]
+        const thirdPoly =
+              variables[2]
+              ? query.isMul(variables[2])
+              ? variables[2].args
+          : [variables[2]]
+        : null
```

Comment:
The indentation is a little off here.  It should probably be:
```
const thirdPoly =
    variables[2]
       ? query.isMul(variables[2])
            ? variables[2].args 
            : [variables[2]]
       : null
```
even that is hard to parse... since the default is `null` maybe do:
```
let thirdPoly = null;
if (variables[2]) {
   if (query.isMul(variables[2]) {
      thirdPoly = variables[2].args
   } else {
      thirdPoly = [variables[2]]
   }
}
```